controllers/replace_failed_process_groups_test.go

```diff
@@ -63,6 +69,595 @@ var _ = Describe("replace_failed_process_groups", func() {
+			}).WithTimeout(time.Duration(cluster.GetTaintReplacementTimeSeconds()*3) * time.Second).WithPolling(1 * time.Second).Should(BeNil())
+
+			Expect(getRemovedProcessGroupIDs(cluster)).To(Equal([]fdbv1beta2.ProcessGroupID{}))
+		})
+
+		When("multiple nodes are tainted", func() {
+			var taintedNodes []*corev1.Node
+			var setValidTaint bool
+			BeforeEach(func() {
+				concurrentTaints := 2
+				Expect(len(allPods)).To(BeNumerically(">", concurrentTaints))
+				taintedNodesIndex := map[int]struct{}{}
+				taintedNodes = []*corev1.Node{}
+				var taintKey string
+				var taintTimeAdded *metav1.Time
+				for len(taintedNodesIndex) < concurrentTaints {
+					taintedNodesIndex[rand.Intn(len(allPods))] = struct{}{}
+				}
+				for key := range taintedNodesIndex {
+					curPod := allPods[key]
+					curNode := &corev1.Node{
+						ObjectMeta: metav1.ObjectMeta{Name: curPod.Spec.NodeName},
+					}
+					taintedNodes = append(taintedNodes, curNode)
+				}
+				for i, taintedNode := range taintedNodes {
+					if i%2 == 0 {
+						taintKey = ""
+						taintTimeAdded = &metav1.Time{Time: time.Now()}
+					} else {
+						taintKey = taintKeyMaintenance
+						taintTimeAdded = nil
+					}
+					if setValidTaint {
+						taintKey = taintKeyMaintenance
+						taintTimeAdded = &metav1.Time{Time: time.Now()}
+					}
+					taintedNode.Spec.Taints = []corev1.Taint{
+						{
+							Key:       taintKey,
+							Value:     "rack_maintenance",
+							Effect:    corev1.TaintEffectNoExecute,
+							TimeAdded: taintTimeAdded,
+						},
+					}
+
+					Expect(k8sClient.Update(ctx.TODO(), taintedNode)).NotTo(HaveOccurred())
+					log.Info("Taint node", "Index", i, "Node name", taintedNode.Name, "Node taints", taintedNode.Spec.Taints)
+				}
+				// Replace all tainted nodes in one reconciliation loop
+				cluster.Spec.AutomationOptions.MaxConcurrentReplacements = &concurrentTaints
+				Expect(k8sClient.Update(ctx.TODO(), cluster)).NotTo(HaveOccurred())
+
+				time.Sleep(time.Second * time.Duration(*cluster.Spec.AutomationOptions.Replacements.TaintReplacementTimeSeconds+1))
+			})
+
+			It("should remove all pods on tainted nodes", func() {
+				retry := len(taintedNodes) * 2 // Hack: Ensure reconciler replaces all tainted nodes
+				for {                          // re-run reconcileCluster up to retry times, assuming each reconciliation replaces at least one pod
+					result, err := reconcileCluster(cluster)
+					Expect(err).NotTo(HaveOccurred())
+					if !result.Requeue || retry <= 0 {
+						break
+					}
+					time.Sleep(time.Microsecond * time.Duration(500)) // Removing this will cause test failure because not all tainted pods are removed
+					retry = retry - 1
+				}
+
+				Expect(getRemovedProcessGroupIDs(cluster)).To(Equal([]fdbv1beta2.ProcessGroupID{}))
+			})
+
+			It("should not remove pods on tainted nodes whose taint Key or TimeAdded is not set", func() {
+				retry := len(taintedNodes) * 2 // Hack: Ensure reconciler replaces all tainted nodes
+				for {
```

Comment:
Instead of using a for loop here and building our own `Eventually` statement we should just use `Eventually`.

Reply:
the test case is removed